File: native/android/app/src/main/java/com/smartmaps/ae/ui/JourneyPlannerScreen.kt
package com.smartmaps.ae.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.smartmaps.ae.core.AECore
import com.smartmaps.ae.core.GeoPoint

/**
 * Journey planner — start → destination over the offline road graph.
 * Start defaults to the live fused position (automatic); destination accepts
 * "lat, lng" today, with place search landing when a geocoder module exists
 * (it activates then — nothing here waits for it).
 */
@Composable
fun JourneyPlannerScreen(core: AECore, onRoutePlanned: () -> Unit) {
    val fix by core.fix.collectAsState()
    val route by core.route.collectAsState()
    var destText by remember { mutableStateOf("") }
    var error by remember { mutableStateOf<String?>(null) }

    Column(Modifier.fillMaxSize().padding(20.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text("Journey planner", style = MaterialTheme.typography.headlineSmall)

        Text(
            "From: " + (fix?.let { "%.5f, %.5f (current position)".format(it.point.lat, it.point.lng) }
                ?: "acquiring position…"),
            style = MaterialTheme.typography.bodyMedium
        )

        OutlinedTextField(
            value = destText,
            onValueChange = { destText = it; error = null },
            modifier = Modifier.fillMaxWidth(),
            placeholder = { Text("Destination (lat, lng)") },
            singleLine = true,
            isError = error != null,
            supportingText = { error?.let { Text(it) } }
        )

        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            Button(
                onClick = {
                    val dest = parseLatLng(destText)
                    if (dest == null) {
                        error = "Enter as: 51.5074, -0.1278"
                    } else {
                        val planned = core.navigateTo(dest)
                        if (planned == null) error = "No route found on the offline graph"
                        else onRoutePlanned()
                    }
                },
                enabled = fix != null
            ) { Text("Plan route") }

            if (route != null) {
                Button(onClick = { core.clearRoute() }) { Text("Clear") }
            }
        }

        route?.let { r ->
            Text(
                "${formatDistance(r.distanceM)}  •  ${formatEta(r.etaS)}",
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.primary
            )
            LazyColumn(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                items(r.steps) { step ->
                    Card(Modifier.fillMaxWidth()) {
                        Column(Modifier.padding(12.dp)) {
                            Text(step.instruction, style = MaterialTheme.typography.bodyLarge)
                            Text(
                                formatDistance(step.distanceM) + " from start",
                                style = MaterialTheme.typography.labelSmall
                            )
                        }
                    }
                }
            }
        }
    }
}

private fun parseLatLng(text: String): GeoPoint? {
    val parts = text.split(",").map { it.trim() }
    if (parts.size != 2) return null
    val lat = parts[0].toDoubleOrNull() ?: return null
    val lng = parts[1].toDoubleOrNull() ?: return null
    if (lat !in -90.0..90.0 || lng !in -180.0..180.0) return null
    return GeoPoint(lat, lng)
}

File: native/ios/SmartMapsAE/UI/JourneyPlannerScreen.swift
// JourneyPlannerScreen.swift — start → destination over the offline road
// graph. Start defaults to the live fused position (automatic); destination
// accepts "lat, lng" today, with place search activating when a geocoder
// module exists.

import SwiftUI

struct JourneyPlannerScreen: View {
    @ObservedObject var core: AECore
    let onRoutePlanned: () -> Void

    @State private var destText = ""
    @State private var error: String?

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Journey planner").font(.title2.bold())

            Text(fromLabel).font(.subheadline)

            TextField("Destination (lat, lng)", text: $destText)
                .textFieldStyle(.roundedBorder)
                .onChange(of: destText) { _ in error = nil }
            if let error {
                Text(error).font(.caption).foregroundColor(.red)
            }

            HStack(spacing: 12) {
                Button("Plan route") { plan() }
                    .buttonStyle(.borderedProminent)
                    .disabled(core.fix == nil)
                if core.route != nil {
                    Button("Clear") { core.clearRoute() }
                        .buttonStyle(.bordered)
                }
            }

            if let route = core.route {
                Text("\(formatDistance(route.distanceM))  •  \(formatEta(route.etaS))")
                    .font(.headline)
                    .foregroundColor(.blue)
                List(route.steps.indices, id: \.self) { i in
                    let step = route.steps[i]
                    VStack(alignment: .leading, spacing: 2) {
                        Text(step.instruction).font(.body)
                        Text("\(formatDistance(step.distanceM)) from start")
                            .font(.caption2)
                            .foregroundColor(.secondary)
                    }
                }
                .listStyle(.plain)
            }

            Spacer()
        }
        .padding(20)
    }

    private var fromLabel: String {
        if let fix = core.fix {
            return String(format: "From: %.5f, %.5f (current position)", fix.point.lat, fix.point.lng)
        }
        return "From: acquiring position…"
    }

    private func plan() {
        guard let dest = parseLatLng(destText) else {
            error = "Enter as: 51.5074, -0.1278"
            return
        }
        if core.navigateTo(dest) == nil {
            error = "No route found on the offline graph"
        } else {
            onRoutePlanned()
        }
    }

    private func parseLatLng(_ text: String) -> GeoPoint? {
        let parts = text.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
        guard parts.count == 2,
              let lat = Double(parts[0]), let lng = Double(parts[1]),
              (-90...90).contains(lat), (-180...180).contains(lng) else { return nil }
        return GeoPoint(lat: lat, lng: lng)
    }
}
